lib/src/history/code_history_controller.dart

```diff
@@ -46,45 +52,109 @@ class CodeHistoryController {
     if (isTextChanging) {
+      _wasTextChanged = true;
       _dropRedoIfAny();
     }
 
-    bool shouldSave = false;
+    final action = _getAction(
+      code: code,
+      selection: selection,
+      isTextChanging: isTextChanging,
+    );
 
-    if (!shouldSave && _wasTextChanged) {
-      // Inserting and deleting lines are significant enough
-      // to save a record without waiting for idle.
-      shouldSave = code.lines.lines.length != lastCode.lines.lines.length;
+    switch (action) {
+      case HistoryControllerAction.newEntriesBeforeAndAfterEdit:
+        _push();
+        lastCode = code;
+        lastSelection = selection;
+        _push();
+        break;
+
+      case HistoryControllerAction.newEntryAfterEdit:
+        lastCode = code;
+        lastSelection = selection;
+        _push();
+        break;
+
+      case HistoryControllerAction.setTimer:
+        _setTimer();
+        break;
     }
 
-    if (!shouldSave) {
-      if (isTextChanging) {
-        _wasTextChanged = true;
+    lastCode = code;
+    lastSelection = selection;
+
+    _removeLastRedundantSelectionOnlyChanges();
+  }
+
+  HistoryControllerAction _getAction({
+    required Code code,
+    required TextSelection selection,
+    required bool isTextChanging,
+  }) {
+    {
+      // If the line count is changed, we add record before and after edit.
+      if (code.lines.length != lastCode.lines.length) {
+        return HistoryControllerAction.newEntriesBeforeAndAfterEdit;
       }
+    }
 
-      if (_wasTextChanged) {
-        final isText1CharLonger = code.text.length == lastCode.text.length + 1;
-        final isTypingContinuous = isText1CharLonger &&
-            selection.hasMovedOneCharacterRight(lastSelection);
+    {
+      // If change is caused by casual typing, we re-set the timer.
+      final isText1CharLonger = code.text.length == lastCode.text.length + 1;
+      final isTypingContinuous = isText1CharLonger &&
+          selection.hasMovedOneCharacterRight(lastSelection);
+      if (isTypingContinuous) {
+        return HistoryControllerAction.setTimer;
+      }
+    }
 
-        if (isTypingContinuous) {
-          _setTimer();
-        } else {
-          shouldSave = true;
+    // Any other change will create record after edit.
+    return HistoryControllerAction.newEntryAfterEdit;
+  }
+
+  void _removeLastRedundantSelectionOnlyChanges() {
+    switch (stack.length) {
+      case 0:
+      case 1:
+        return;
+
+      case 2:
+        // Drop the first one if only selection has changed.
+        if (_isFullTextSame([stack[0], stack[1]])) {
+          stack.removeAt(0);
+        }
+        break;
+
+      default:
+        // Check last, last-1, last-2. Drop last-1.
+        final last = stack.last;
+        final lastMinus1 = stack[stack.length - 2];
+        final lastMinus2 = stack[stack.length - 3];
+
+        if (_isFullTextSame([last, lastMinus1, lastMinus2])) {
+          stack.removeAt(stack.length - 2);
         }
```

Comment:
Also check if `lastMinus2 == lastMinus3`. This happens with multiple `enter` presses. Remove the check from `_push()`.